### PR TITLE
[Unicode] support startswith with args, start and end.

### PR DESCRIPTION
**Pull request by dlee992**
As it is, support `start` and `end` optional arguments for `startswith`.

Most added code comes from the implementation of `endswith`, with marginal changes.

PS, I also have aother PR about supporting `int(str)` based on this one, and I will open that one after merging this.

BTW, I have a question about `_nrt=False`, I didn't find any document about this, what is it for in the Numba Runtime level? Disable any memory allocation/refct/release?

### CODE REVIEW

File: numba/cpython/unicode.py
Comment (dlee992) [reply numba/cpython/unicode.py]: No problem. Added, wait for CI.

**Review by guilhermeleobas — CHANGES_REQUESTED** [review]

**Comment by guilhermeleobas** [comment]
I have checked that:

- [x]  There is no other Pull Request proposing the same change. (If there is
please highlight this and one of the core developers will pick it up.)
- [ ]  The proposed additional NumPy functionality is not deprecated by
NumPy.
- [x]  Public CI is passing. (With the exception of something unrelated to this PR)
- [x]  The patch does not touch files in `numba.core`.
- [x]  The patch is implemented using the high-level extension API only
([https://numba.readthedocs.io/en/stable/extending/high-level.html](https://numba.readthedocs.io/en/stable/extending/high-level.html)).
- [ ]  The patch updates the documentation to reflect the new functionality.
([https://github.com/numba/numba/blob/main/docs/source/reference/numpysupported.rst](https://github.com/numba/numba/blob/main/docs/source/reference/numpysupported.rst))
- [x]  The patch is line wrapped to 80 characters.
- [x]  Changes to Python source in the patch are `flake8` compliant.
- [x]  The unit tests for the implementation:
    - [x]  Explicitly check custom error messages are raised appropriately.
    - [x]  Hit all the branches of the proprosed change (both at type
    inference and run time).
- [x]  The patch works in a simple manual test in my local checkout.
- [x]  I cannot break the code in the patch in my local checkout after
trying to do so for a minimum of 10 minutes. The error messages seem
reasonable and there are no "incorrect" results.
- [x]  I think the code introduced in this change is easy to maintain.
- [x]  I understand all the code in this PR.
- [ ]  The implementation of the NumPy/CPython function reflects that present in the
NumPy implementation itself (look in the NumPy repository:
[https://github.com/numpy/numpy](https://github.com/numpy/numpy))
    - [ ]  The implementation has references to the region of the NumPy code
    base that provides the same functionality. The references are URLs
    of the form:
    `https://github.com/numpy/numpy/tree/<SHA>/path/to/file.py:#<lines>`
    - [ ]  The code "looks" reasonably similar to the NumPy/CPython implementation.
    - [ ]  The code uses the same algorithm as the NumPy/CPython implementation, if
    it does not it explains why and this justification is reasonable.
    - [x]  The type inference part of the Numba implementation restricts the
    types to those declared in the NumPy/CPython docs (or if not present, what
    is accepted in the NumPy/CPython implementation itself).
    - [x]  The Numba implementation uses the same variable names for
    arguments as in the NumPy/CPython implementation.
    - [ ]  The Numba code has commentary present for anything not-obvious by
    inspection.
    - [ ]  The Numba code accommodates differences that might occur across
    different NumPy versions in a clear fashion and uses
    `numba.np.numpy_support.numpy_version` to separate the difference
    in implementations.
- [x]  The unit tests for the implementation:
    - [x]  Use the same test inputs as the NumPy/CPython unit tests for the same
    functionality. If this is not the case there are comments which
    explain why and this justification is reasonable.
    - [ ]  The NumPy/CPython unit test inputs have reference URLs associated with
    them to track their origin. The references are URLs of the form:
    `https://github.com/numpy/numpy/tree/<SHA>/path/to/file.py:#<lines>`
    - [x]  Use `unittest` style assertions (not pytest style!).
    - [x]  Check that unsupported/invalid input types are handled as expected
    using the `with self.assertRaises()` pattern.
    - [ ]  Check that numerically awkward values work as expected e.g.
    `NaN`, `+/-Inf`, `0`, `INT_MAX`, `INT_MIN` etc.
    - [ ]  Check that types which can be converted to arrays (assuming NumPy
    accepts this behaviour) work as expected.
    - [ ]  Check that passing empty versions of containers e.g. empty tuple
    or 0d array works as expected.

File: numba/cpython/unicode.py
Comment (dlee992) [reply numba/cpython/unicode.py:834]: Sure. Keeping argument names consistent is a good habit.

File: numba/tests/test_unicode.py
Comment (dlee992) [reply numba/tests/test_unicode.py]: Okay. Revised as it should be.

**Comment by dlee992** [comment]
Actually, the API `endswith` implementation contains similar "nitpicks" as my previous version for `startswith`. But I guess we could change that later, after this one is merged.

**Review by guilhermeleobas — APPROVED** [review]

File: numba/cpython/unicode.py
Comment (stuartarchibald) [inline numba/cpython/unicode.py:838]: I'm not sure this error message is correct? `None` is accepted so it would be entirely legal for a user to write:
```python
'abcdef'.startswith('f', None)
```
which is not an integer.

File: numba/cpython/unicode.py
Comment (stuartarchibald) [inline numba/cpython/unicode.py:843]: Same comment as above.

File: numba/cpython/unicode.py
Comment (stuartarchibald) [inline numba/cpython/unicode.py:835]: Should this use `numba.core.cgutils.is_nonelike` oppose to both checking for `None`, `types.NoneType` and `types.Omitted`?

File: numba/cpython/unicode.py
Comment (stuartarchibald) [inline numba/cpython/unicode.py:845]: `Tuple` cannot be accepted here unless the contents of the `Tuple` are all `types.StringLiteral`. Consider:

```python
@njit
def foo():
    print('abcde'.startswith(('a', 1)))

foo()
```
it may be easier to just not accept `Tuple` to force type inference to try with a non-literal types instead.

File: numba/cpython/unicode.py
Comment (stuartarchibald) [inline numba/cpython/unicode.py:878]:
```suggestion
            s_slice = s[start:end]

            return _cmp_region(s_slice, 0, prefix, 0, sub_length) == 0
```
just to make it clear that this is addressing a slice of the original string.

**Review by stuartarchibald — CHANGES_REQUESTED** [review]
Thanks for the patch. Few minor things to look at else looks good.

File: numba/cpython/unicode.py
Comment (dlee992) [reply numba/cpython/unicode.py:845]: I prefer to accept `Tuple` of `StringLiteral`, 1) it's more consistent with cpython, 2) I do need this to simplify my next PR https://github.com/numba/numba/pull/8558. So I prefer to add a check to make sure every item in `Tuple` (or `UniTuple`) is a `StringLiteral`.

File: numba/cpython/unicode.py
Comment (dlee992) [reply numba/cpython/unicode.py:838]: Revised as expected.

File: numba/cpython/unicode.py
Comment (dlee992) [reply numba/cpython/unicode.py:835]: Revised using this func.

File: numba/cpython/unicode.py
Comment (dlee992) [reply numba/cpython/unicode.py:845]: Revised, please review this issue again.

File: numba/cpython/unicode.py
Comment (dlee992) [reply numba/cpython/unicode.py:878]: OK.

File: numba/cpython/unicode.py
Comment (stuartarchibald) [inline numba/cpython/unicode.py:848]: Just accept `types.UniTuple` with a `.dtype` which is `types.UnicodeType` ? e.g.

```python
if isinstance(prefix, types.UniTuple) and isinstance(prefix.dtype, types.UnicodeType):
    ... # then return the implementation
```

File: numba/cpython/unicode.py
Comment (stuartarchibald) [inline numba/cpython/unicode.py:858]:
```suggestion
    elif isinstance(prefix, types.UnicodeCharSeq):
```

File: numba/cpython/unicode.py
Comment (stuartarchibald) [inline numba/cpython/unicode.py:864]:
```suggestion
    elif isinstance(prefix, types.UnicodeType):
```

File: numba/cpython/unicode.py
Comment (stuartarchibald) [inline numba/cpython/unicode.py:884]: I think there should be an `else` here for the case where `prefix` is of an unsupported type, which raises:
```python
    else:
        raise TypingError("The arg 'prefix' should be a string or a tuple of strings")
```

**Review by stuartarchibald — COMMENTED** [review]
Thanks for the update @dlee992. I've made a few suggestions about restructuring the type checking with respect to the type of `prefix` (and a few other minor things). I think once resolved this should be ready to merge. 

Note: I'll answer your questions on the diff inline now so as to not get them confused with this review.

File: numba/cpython/unicode.py
Comment (stuartarchibald) [reply numba/cpython/unicode.py]: By default, type inference will try using non-literal types first, and then if that fails, it will try again using literal types. Python source containing tuples of string literals will be inferred as `types.Tuple(<types.StringLiteral>, <types.StringLiteral>, ...)`. e.g.
```python
from numba import njit
import numpy as np

@njit
def foo():
    x = ('a', 'b')

foo()
foo.inspect_types()
```
-------------------------------------
You can see the "try non-literal then literal" types behaviour by doing something like:
```python
from numba import njit, types
from numba.extending import overload
import numpy as np

def bar(tup):
    pass

@overload(bar)
def ol_bar(tup):
    print(f"bar:: tup type is {tup}")

@njit
def foo():
    x = ('a', 'b')
    bar(x)

foo()
```
(ignore the compilation error, just look at what gets printed to stdout).

-------------------------------------
RE: the `'a'.startswith(('a', 5))`, there are implementation details like this in the CPython implementation. They don't need replicating unless there's some proven use case!

**Review by stuartarchibald — APPROVED** [review]
Thanks for the patch and fixes!

**Pull request merged**
